### PR TITLE
open and close obj for nest + new tests

### PR DESCRIPTION
**Pull request by Amine-Mike**

### CODE REVIEW

File: test/customization_test.rb [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  end
+end
+
+describe 'Line delimiter tests' do
+  it 'must work with a \n delimiter' do
+    printer = Oppen::Wadler.new(margin: 45, new_line: '\n')
+    customization_build_output(printer)
+    _(printer.output)
+      .must_equal 'function\n  test(\n    int index, char character,\n    float precision, double trouble\n  )'
+  end
```
Comment (stackmystack) [inline test/customization_test.rb]: Nah, don't use `\n`. It's confusing. Use a visible ASCII char.

Actually, now that I think about it, we should also add tests on UTF-8 chars outside the ASCII range.

File: test/customization_test.rb [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    nest_delimiter_build_output(printer, '', '}')
+    _(printer.output).must_equal <<~LANG.chomp
+      function
+        foo()
+      Hello
+        World!
+      }
+    LANG
+  end
+end
```
Comment (stackmystack) [inline test/customization_test.rb]: We should also add a test on `nil` in left and right.

But should we raise an exception here, or simply coecre it to an empty space?
I personally dislike magical `nil`s …

File: lib/wadler/print.rb [hunk 3]
```diff
@@ -75,17 +75,28 @@ def group(indent = 0, open_obj = '', close_obj = '',
     end
 
     # @param indent [Integer] nest indentation
+    # @param open_obj [String] nest oppening delimiter
```
Comment (akimd) [inline lib/wadler/print.rb]: Pun intended?

Comment (Amine-Mike) [reply lib/wadler/print.rb]: I don't see the pun...

Comment (akimd) [reply lib/wadler/print.rb]: nest **oppening** delimiter

File: lib/wadler/print.rb [hunk 4]
```diff
@@ -75,17 +75,28 @@ def group(indent = 0, open_obj = '', close_obj = '',
...
+    # @param close_obj [String] nest closing delimiter
     # @param break_type [Oppen::Token::BreakType] nest breaking type
     #
     # @return [Nil]
-    def nest(indent, break_type = Oppen::Token::BreakType::CONSISTENT, &block)
+    def nest(indent, open_obj = '', close_obj = '',
+             break_type = Oppen::Token::BreakType::CONSISTENT, &)
```
Comment (akimd) [inline lib/wadler/print.rb]: When using yield, one usually does not even mention the block at all.

Besides, make sure all this works in Ruby 3.1, which is, AFAICT, the current env in prod.

File: lib/wadler/print.rb [hunk 5]
```diff
@@ -51,6 +51,9 @@ def output
     # @return [Nil]
     def group(indent = 0, open_obj = '', close_obj = '',
               break_type = Oppen::Token::BreakType::CONSISTENT, &)
+      raise ArgumentError, "#{open_obj.nil? ? 'open_obj' : 'close_obj'} cannot be nil" \
+      if open_obj.nil? || close_obj.nil?
+
```
Comment (akimd) [inline lib/wadler/print.rb]: This is hideous.  Either:
- intend the continued line
- keep it on one long line
- use a regular if.

File: lib/wadler/print.rb [hunk 6]
```diff
@@ -93,7 +99,7 @@ def nest(indent, open_obj = '', close_obj = '',
     ensure
       @current_indent -= indent
 
-      if !close_obj.empty?
+      if !close_obj&.empty?
```
Comment (akimd) [inline lib/wadler/print.rb]: Why don't use the straightforward `if close_obj`?  I very much doubt Booleans are accepted here anyway.  Keep it idiomatic.

Comment (akimd) [reply lib/wadler/print.rb]: I meant `if close_obj.present?`, my bad.  But given the code, I see that anyway nil is invalid, so I don't understand why you try to support nil here.

File: lib/wadler/print.rb (hunk 3 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb]: Oh, didn't even realize. It appears that the typo checker also didn't notice. Will fix this

File: lib/wadler/print.rb (hunk 4 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb]: The CI that Firas setup runs the tests on ruby 3.2 and 3.3. Should I add version 3.1?

File: lib/wadler/print.rb (hunk 6 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb]: This portion of the code is present in the ensure block. If the method receives a nil close_obj and I don't check for nil, the program breaks (accessing nil element) because the ensure block will be executed after I raise the ArgumentError

File: lib/wadler/print.rb (hunk 5 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb]: I used this syntax in the past, in lib/oppen/print_stack.rb: 49. Do you want me to also change it there?

File: lib/wadler/print.rb (hunk 6 above)
Comment (akimd) [reply lib/wadler/print.rb]: So maybe you should rather use a nested `begin/ensure/end` block inside the routine, and check preconditions **before**.

As it is, it looks clumsy to me.

Actually, anyway, move all precondition checks before any actual content (not even the `@current_indent thing`.

See what I mean?

Comment (Amine-Mike) [reply lib/wadler/print.rb]: yes

File: lib/wadler/print.rb (hunk 5 above)
Comment (akimd) [reply lib/wadler/print.rb]: Yes.

File: lib/wadler/print.rb (hunk 4 above)
Comment (akimd) [reply lib/wadler/print.rb]: Yes.

File: lib/wadler/print.rb [hunk 7]
```diff
@@ -75,17 +78,33 @@ def group(indent = 0, open_obj = '', close_obj = '',
...
+        @current_indent -= indent
+      end
+
+      return if close_obj.empty?
 
-      block.call
-    ensure
-      @current_indent -= indent
+      self.break
+      text(close_obj)
```
Comment (akimd) [inline lib/wadler/print.rb:107]: I would move the whole "true" content in the begin/end.  And avoid return (but in error handling).

```ruby
    def nest(indent, open_obj = '', close_obj = '',
              break_type = Oppen::Token::BreakType::CONSISTENT)
       raise ArgumentError, "#{open_obj.nil? ? 'open_obj' : 'close_obj'} cannot be nil" \
         if open_obj.nil? || close_obj.nil?

       begin
         @current_indent += indent

         if !open_obj.empty?
           text(open_obj)
           self.break
         end

         yield

         if !close_obj.empty?
           self.break
           text(close_obj)
         end
       ensure
         @current_indent -= indent
       end
     end
```

File: lib/wadler/print.rb (hunk 4 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb]: After verification, the README.md in langeod specifies that ruby 3.2+ is required. No need to add version 3.1 to the CI

File: lib/wadler/print.rb (hunk 7 above)
Comment (Amine-Mike) [reply lib/wadler/print.rb:107]: This doesn't work, we need to decrement current_indent before displaying the close_obj or else the we will not get the desired output.

**Review by akimd — APPROVED** [review]
